pigpen-avro/src/main/clojure/pigpen/avro/core.clj

```diff
@@ -0,0 +1,120 @@
+                 (= type Schema$Type/UNION)  (let [record-fields (->> (.getTypes (.schema field))
+                                                                      (filter #(= (.getType ^{:tag Schema} %) Schema$Type/RECORD))
+                                                                      (map (fn [rec] (prefixed-names rec (.name field)))))]
+                                               (if (empty? record-fields) [(.name field)]
+                                                   record-fields))
+                 :else                       (.name field)))))
+       flatten
+       vec))
+
+(defn parse-schema [^String s] (let [parser ^Schema$Parser (Schema$Parser.)]
+                         (.parse parser s)))
+
+(defmethod pigpen.pig/native->clojure org.apache.avro.util.Utf8 [value]
+  (str value))
+
+(defn dotted-keys->nested-map [kvs]
+  (->> kvs
+   (map (fn [[k v]] [(-> k name (clojure.string/split #"\.")) v]))
+   (reduce (fn [acc [ks v]]
+             (try
+              (if v (assoc-in acc (map keyword ks) v) acc)
+              (catch Exception e (throw (Exception. (str "can't assoc-in record: " acc "\nkeys: " ks "\nvalue: " v "\nkvs: " kvs)))))) {})))
+
+
+(defn load-avro
+  "*** ALPHA - Subject to change ***
+
+  Loads data from an avro file. Returns data as maps with keyword keys corresponding to avro field names. Fields with avro type \"map\" will be maps with string keys.
+
+  Example:
+
+    ;; avro schemas are defined [on the project's website](http://avro.apache.org/docs/1.7.7/spec.html#schemas)
+    ;; load-avro takes the schema as a string.
+    (pig-avro/load-avro \"input.avro\" (slurp \"schemafile.json\"))
+
+    (pig-avro/load-avro \"input.avro\"
+       \"{\"namespace\": \"example.avro\", \"type\": \"record\", \"name\": \"foo\",
+                       \"fields\": [{\"name\": \"wurdz\", \"type\": \"string\"}, {\"name\": \"bar\", \"type\": \"int\"}] }\")
+
+"
+  ([location schema]
+     (let [^Schema parsed-schema (parse-schema schema)
+           storage (raw/storage$
+                    ["piggybank.jar"]
+                    "org.apache.pig.piggybank.storage.avro.AvroStorage"
+                    ["schema" (SchemaNormalization/toParsingForm parsed-schema)])
+           field-symbols (map symbol (field-names parsed-schema))]
+       (->
+        (raw/load$ location field-symbols storage {:implicit-schema true})
+        (raw/bind$
+         '[pigpen.avro.core]
+         '(pigpen.pig/map->bind (comp
+                                 pigpen.avro.core/dotted-keys->nested-map
+                                 (pigpen.pig/args->map pigpen.pig/native->clojure)))
+         {:args (clojure.core/mapcat (juxt str identity) field-symbols)
+          :field-type-in :native })))))
+
+(defn attrs-lookup [^GenericData$Record record attr-names]
+  (cond
+   (nil? record) nil
+   (empty? attr-names) record
+   :else (attrs-lookup (.get record ^{:tag java.lang.String} (first attr-names)) (rest attr-names))))
```

Comment:
Theoretically they can be as nested as you like. In practice it would be a very strange schema that was nested enough to worry about how much stack we're using.